crates/starknet/src/starknet/mod.rs

```diff
@@ -576,6 +580,114 @@ impl Starknet {
+
+        let simulated = simulate_transaction(
+            simulatable.as_slice(),
+            state.state.clone(),
+            &self.block_context,
+            INITIAL_SIMULATION_GAS,
+            skip_validate,
+            skip_execute,
+            skip_fee_charge,
+            ignore_max_fee,
+            skip_nonce_check,
+        )?;
+
+        let estimated = self.estimate_fee(block_id, transactions)?;
+
+        // if the underlying simulation is correct, this should never be the case
+        // in alignment with always avoiding assertions in production code, this has to be done
+        if simulated.len() != estimated.len() {
+            return Err(Error::UnexpectedInternalError {
+                msg: format!(
+                    "Non-matching number of simulations ({}) and estimations ({})",
+                    simulated.len(),
+                    estimated.len()
+                ),
+            });
+        }
+
+        let mut simulation_results: Vec<SimulatedTransaction> = vec![];
+        for (tx_execution_info, fee_estimation) in simulated.into_iter().zip(estimated) {
+            let validate_invocation = if let Some(validate_info) = tx_execution_info.validate_info {
+                Some(validate_info.try_into()?)
+            } else {
+                None
+            };
+
+            let fee_transfer_invocation =
+                if let Some(fee_transfer_info) = tx_execution_info.fee_transfer_info {
+                    Some(fee_transfer_info.try_into()?)
+                } else {
+                    None
+                };
+
+            let transaction_trace = match tx_execution_info.tx_type {
+                Some(starknet_in_rust::definitions::transaction_type::TransactionType::Declare) => TransactionTrace::Declare(DeclareTransactionTrace {
+                        validate_invocation,
+                        fee_transfer_invocation,
+                    }),
+                Some(starknet_in_rust::definitions::transaction_type::TransactionType::DeployAccount) => TransactionTrace::DeployAccount(DeployAccountTransactionTrace {
+                    validate_invocation,
+                    constructor_invocation: if let Some(call_info) = tx_execution_info.call_info {
+                        Some(call_info.try_into()?)
+                    } else {
+                        None
+                    },
+                    fee_transfer_invocation,
+                }),
+                Some(starknet_in_rust::definitions::transaction_type::TransactionType::InvokeFunction) => {
+                    let call_info = tx_execution_info.call_info.clone();
+                    let success = call_info.is_some() && call_info.unwrap().result().is_success;
+                    TransactionTrace::Invoke(InvokeTransactionTrace {
+                        validate_invocation,
+                        execution_invocation: match success {
+                            // safe to unwrap, checked above
```

Comment:
even if it's checked shouldn't we use match and avoid unwrap?

Reply:
Removed the .unwrap in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/189/commits/15310d5fa1ea32a879d55e1a226999c28bd94b38